File: Easy/Container_With_Most_Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # first attempt: don't brute force it, instead figure out a way to 
        # find what is most likely the largest and then go from there
        # so check largest bar with furthest bar and go from there

        max = 0
        left = 0
        right = len(height) - 1
        max = min((height[left], height[right])) * (right - left)
        for i in range(len(height)):
            if(min((height[left], height[right])) * (right - left) > max):
                max = min((height[left], height[right])) * (right - left)
            if(height[left] < height[right]):
                left = left + 1
            else:
                right = right - 1
      

        return max

File: Medium/Container_With_Most_Water.java
class Solution {
    public int maxArea(int[] height) {
        // brute force is O(n^2) to check every combination
        // optimal is O(n) where you use left and right

        int left = 0;
        int right = height.length-1;
        int maxArea = 0;
        while(left < right){
            if (maxArea < ( (right - left) * Math.min(height[right],height[left]) )){
                maxArea = (right - left) * Math.min(height[right],height[left]);
               
            }
            if(height[left] < height[right]){
                left++;
            }
            else if(height[left] > height[right]){
                right--;
            }
            else{
                if(height[left+1] < height[right-1]){
                    right--;
                }
                else if(height[left+1] > height[right-1]){
                    left++;
                }
                else{
                    left++;
                }
            }
            
        }


        return maxArea;
    }
}